### PR TITLE
PBftState: deserialise without `k`

### LINKED ISSUE #2383: Avoid need for `k` in `decodeByronChainDepState`
Right now this means `k` is part of the `CodecConfig`, which means client code needs to be aware of it. We could just store `k` alongside the rest of the state (at the cost of not being able to check that it's the right value).

We should probably do this as part of #2156 .

### PR DESCRIPTION
**Pull request by mrBliss**
Fixes #2383.

Old format:
  1. An anchor (`WithOrigin SlotNo`)
  2. A list of `PBftSigner`s
  3. `MaybeEbbInfo`

The four `PBftSigner` sequences can be reconstructed from 1, 2, the
`SecurityParam` and `WindowSize`. The reliance on these last two parameters is
problematic and requires the `SecurityParam` in the `CodecConfig` for Byron (and
mock PBft), which we would prefer to avoid. The end goal is to have an empty
`CodecConfig`, like for Shelley.

Moreover, this old format was rather wasteful: while there are only 7 different
`PBftVerKeyHash`s, we end up storing 2160 of them.

New format:
  1. `Map (PBftVerKeyHash c) ([SlotNo], [SlotNo], [SlotNo], [SlotNo])`
     For each signer: the slots in `preAnchor, postAnchor`, `preWindow`, and
     `inWindow.
  2. `MaybeEbbInfo`

Now we no longer need the `SecurityParam` and `WindowSize` to decode a
`PBftState`. Bonus: we only store each `PBftVerKeyHash` once.

Cost: we store more slots than before, but to avoid that, we'd have to store the
`WindowSize` on disk. That would require either passing it to the encoder (what
we're trying avoid in the first place) or storing it in `PBftState`.

Using the final Byron ledger snapshot to compare the sizes:
Before: 155,534 bytes
After:   43,477 bytes
The new format is 70% smaller than the old.

Unfortunately, this is just a drop in the bucket, because the size of the entire
ledger snapshot drops from 82,633,323 to 82,521,266 bytes, so in the end it
saves not even a percent.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/State.hs [hunk 1]
```diff
@@ -577,44 +579,98 @@ fromList k n (anchor, signers, ebbs) =
...
+    postAnchor = dlToSeq dlPostA
+    preWindow  = dlToSeq dlPreW
+    inWindow   = dlToSeq dlInW
+
+    counts = computeCounts inWindow
+
+encodePBftState :: (PBftCrypto c, Serialise (PBftVerKeyHash c))
                 => PBftState c -> Encoding
-encodePBftState st@PBftState{..} =
-    encodeVersion serializationFormatVersion0 $ mconcat [
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/State.hs:586]: As mentioned in https://github.com/input-output-hk/ouroboros-network/issues/2383#issuecomment-672892991, I'm no longer bothering with `Ouroboros.Consensus.Util.Versioned`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/State.hs [hunk 2]
```diff
@@ -577,44 +579,98 @@ fromList k n (anchor, signers, ebbs) =
...
-
-encodePBftState :: Serialise (PBftVerKeyHash c)
+toCompactFormat ::
+     forall c. PBftCrypto c
+  => PBftState c
+  -> ( Map (PBftVerKeyHash c) ([SlotNo], [SlotNo], [SlotNo], [SlotNo])
+       -- For each signer, the slots in 'preAnchor', 'postAnchor',
+       -- 'preWindow', and 'inWindow'.
+     , MaybeEbbInfo
+     )
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/State.hs:589]: Probably overkill, @edsko will likely say so. I reckon he'd just want me to encode the four `StrictSeq (PBftSigner c)`s directly to keep it simple. But ... bytes :grin:

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/State.hs:589]: I'm so glad you're still having internal dialogues with me even when I'm not around, saves me so much hassle :grimacing:

**Comment by mrBliss** [comment]
#2575 will solve this problem in a much nicer way.

**Pull request closed without merging**